File: rb/test/card_list_entity_test.rb
# CardList entity test

require "minitest/autorun"
require "json"
require_relative "../Scryfall_sdk"
require_relative "runner"

class CardListEntityTest < Minitest::Test
  def test_create_instance
    testsdk = ScryfallSDK.test(nil, nil)
    ent = testsdk.CardList(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = card_list_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "card_list." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set SCRYFALL_TEST_CARD_LIST_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    card_list_ref01_ent = client.CardList(nil)
    card_list_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.card_list"), "card_list_ref01"))

    card_list_ref01_data_result = card_list_ref01_ent.create(card_list_ref01_data, nil)
    card_list_ref01_data = Helpers.to_map(card_list_ref01_data_result)
    assert !card_list_ref01_data.nil?
    assert !card_list_ref01_data["id"].nil?

    # LIST
    card_list_ref01_match = {}

    card_list_ref01_list_result = card_list_ref01_ent.list(card_list_ref01_match, nil)
    assert card_list_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(card_list_ref01_list_result),
      { "id" => card_list_ref01_data["id"] })
    assert !Vs.isempty(found_item)

  end
end

def card_list_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "card_list", "CardListTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = ScryfallSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["card_list01", "card_list02", "card_list03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["SCRYFALL_TEST_CARD_LIST_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "SCRYFALL_TEST_CARD_LIST_ENTID" => idmap,
    "SCRYFALL_TEST_LIVE" => "FALSE",
    "SCRYFALL_TEST_EXPLAIN" => "FALSE",
  })

  idmap_resolved = Helpers.to_map(
    env["SCRYFALL_TEST_CARD_LIST_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["SCRYFALL_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
      },
      extra || {},
    ])
    client = ScryfallSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["SCRYFALL_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["SCRYFALL_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/card_list_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/scryfall-sdk/go"
	"github.com/voxgig-sdk/scryfall-sdk/go/core"

	vs "github.com/voxgig-sdk/scryfall-sdk/go/utility/struct"
)

func TestCardListEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CardList(nil)
		if ent == nil {
			t.Fatal("expected non-nil CardListEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"card_list": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.CardList(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.CardList(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := card_listBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "card_list." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set SCRYFALL_TEST_CARD_LIST_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		cardListRef01Ent := client.CardList(nil)
		cardListRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "card_list"}, setup.data), "card_list_ref01"))

		cardListRef01DataResult, err := cardListRef01Ent.Create(cardListRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		cardListRef01Data = core.ToMapAny(entityData(cardListRef01DataResult))
		if cardListRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if cardListRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		cardListRef01Match := map[string]any{}

		cardListRef01ListResult, err := cardListRef01Ent.List(cardListRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		cardListRef01List, cardListRef01ListOk := cardListRef01ListResult.([]any)
		if !cardListRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", cardListRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(cardListRef01List), map[string]any{"id": cardListRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

	})
}

func card_listBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "card_list", "CardListTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read card_list test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse card_list test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"card_list01", "card_list02", "card_list03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("SCRYFALL_TEST_CARD_LIST_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"SCRYFALL_TEST_CARD_LIST_ENTID": idmap,
		"SCRYFALL_TEST_LIVE":      "FALSE",
		"SCRYFALL_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["SCRYFALL_TEST_CARD_LIST_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["SCRYFALL_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewScryfallSDK(core.ToMapAny(mergedOpts))
	}

	live := env["SCRYFALL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["SCRYFALL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
